tools/clang/lib/SPIRV/SpirvVisitor.cpp

```diff
@@ -0,0 +1,581 @@
+    for (auto fnParam : fn->getParams()) {
+      visit(fnParam);
+    }
+  }
+  // After emitting the function
+  else if (phase == Visitor::Phase::Done) {
+    // Emit OpFunctionEnd
+    initInstruction(spv::Op::OpFunctionEnd, 1);
+    finalizeInstruction();
+  }
+
+  return true;
+}
+
+bool EmitVisitor::visit(SpirvBasicBlock *bb, Phase phase) {
+  assert(bb);
+
+  // Before emitting the basic block.
+  if (phase == Visitor::Phase::Init) {
+    // Emit OpLabel
+    initInstruction(spv::Op::OpLabel, 2);
+    curInst.push_back(getNextId());
+    finalizeInstruction();
+  }
+  // After emitting the basic block
+  else if (phase == Visitor::Phase::Done) {
+    assert(bb->hasTerminator());
+  }
+  return true;
+}
+
+bool EmitVisitor::visit(SpirvCapability *cap) {
+  initInstruction(cap->getopcode(), 2);
+  curInst.push_back(static_cast<uint32_t>(cap->getCapability()));
+  finalizeInstruction();
+  return true;
+}
+
+bool EmitVisitor::visit(SpirvExtension *ext) {
+  initInstruction(ext->getopcode(), 2);
+  encodeString(ext->getExtensionName());
+  finalizeInstruction();
+  return true;
+}
+
+bool EmitVisitor::visit(SpirvExtInstImport *inst) {
+  initInstruction(inst->getopcode(), 3);
+  curInst.push_back(inst->getResultId());
+  encodeString(inst->getExtendedInstSetName());
+  finalizeInstruction();
+  return true;
+}
+
+bool EmitVisitor::visit(SpirvMemoryModel *inst) {
+  initInstruction(inst->getopcode(), 3);
+  curInst.push_back(static_cast<uint32_t>(inst->getAddressingModel()));
+  curInst.push_back(static_cast<uint32_t>(inst->getMemoryModel()));
+  finalizeInstruction();
+  return true;
+}
+
+bool EmitVisitor::visit(SpirvEntryPoint *inst) {
+  initInstruction(inst->getopcode(), 4);
+  curInst.push_back(static_cast<uint32_t>(inst->getExecModel()));
+  curInst.push_back(static_cast<uint32_t>(inst->getEntryPointId()));
+  encodeString(inst->getEntryPointName());
+  curInst.insert(curInst.end(), inst->getInterface().begin(),
+                 inst->getInterface().end());
+  finalizeInstruction();
+  return true;
+}
+
+bool EmitVisitor::visit(SpirvExecutionMode *inst) {
+  initInstruction(inst->getopcode(), 3);
+  curInst.push_back(inst->getEntryPointId());
+  curInst.push_back(static_cast<uint32_t>(inst->getExecutionMode()));
+  curInst.insert(curInst.end(), inst->getParams().begin(),
+                 inst->getParams().end());
+  finalizeInstruction();
+  return true;
+}
+
+bool EmitVisitor::visit(SpirvString *inst) {
+  initInstruction(inst->getopcode(), 3);
+  curInst.push_back(inst->getResultId());
+  encodeString(inst->getString());
+  finalizeInstruction();
+  return true;
+}
+
+bool EmitVisitor::visit(SpirvSource *inst) {
+  initInstruction(inst->getopcode(), 3);
+  curInst.push_back(static_cast<uint32_t>(inst->getSourceLanguage()));
+  curInst.push_back(static_cast<uint32_t>(inst->getVersion()));
+  if (inst->hasFileId())
+    curInst.push_back(inst->getFileId());
+  if (!inst->getSource().empty())
+    encodeString(inst->getSource());
```

Comment:
We are actually copying the whole string two times: one from `encodeSPIRVString()` to `curInst`, the order from `curInst` to `spirvBinary`. It can be costly since we can embed the whole program. Maybe we need to do it in-place in `spirvBinary`. How do you think?

Reply:
Good point. Done.